### PR TITLE
TST: Clean warnings

### PR DESCRIPTION
**Pull request by aburrell**
# Description

Addresses #1009 by adding a general test triggered by an optional testing variable to the standard set of Instrument tests.  Adapted the clean method in the test instruments to run all potential iterations of the allowed behaviour, which includes:

- logging messages
- warnings
- errors
- changing the clean level

Updated the documentation to describe how this testing should be done.

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?

Added unit tests.  Also added this line to `ace_epam.py` in pysatSpaceWeather and ran the local unit tests.

```
# Set clean warning tests
_clean_warn = {'': {'': {'dusty': [('logger', 'WARN',
                         "unused clean level 'dusty', reverting to 'clean'",
                         'clean')]}}}
```

**Test Configuration**:
* Operating system: OS X Big Sur
* Version number: Python 3.9
* Any details about your local setup that are relevant: See also:
    * https://github.com/pysat/pysatSpaceWeather/pull/125 
    * https://github.com/pysat/pysatMadrigal/pull/93
    * https://github.com/pysat/pysatModels/pull/129

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

If this is a release PR, replace the first item of the above checklist with the release
checklist on the wiki: https://github.com/pysat/pysat/wiki/Checklist-for-Release

### CODE REVIEW

File: docs/new_instrument.rst [hunk 1]
```diff
@@ -731,7 +738,25 @@ combinations), and runs the tests using pytestmark.  By default,
...
+or the warning/error type (e.g., ``ValueError``). The third element provides the
+warning message as a string and the final element provides the expected clean
+level after running the clean routine.
 
+.. code:: python
+
+   # ------------------------------------------
+   # Instrument test attributes
+
+   _clean_warn = {'dusty': ('logger', 'WARN', "I am a warning!", 'clean')}
```
Comment (aburrell) [inline docs/new_instrument.rst:756]:
```suggestion
   _clean_warn = {inst_id: {tag: {'dusty': ('logger', 'WARN', "I am a warning!", 'clean')} for tag in inst_ids[inst_id]} for inst_id in inst_ids.keys()}
```

File: pysat/tests/classes/cls_instrument_library.py [hunk 2]
```diff
@@ -366,6 +367,109 @@ def test_load(self, clean_level, inst_dict):
...
+                        "Clean level should now be {:s}, not {:s}".format(
+                            final_level, test_inst.clean_level)
+
+                    # Make sure fake data is cleared
+                    assert target not in test_inst.data
+
+                    # If cleaning not used, something should be in the file. Not
+                    # used for clean levels since cleaning may remove all data
+                    if clean_level == "none":
+                        assert not test_inst.empty
```
Comment (aburrell) [inline pysat/tests/classes/cls_instrument_library.py:469]:
```suggestion
```

These lines can never be reached, since 'none' bypasses the clean method.

File: docs/new_instrument.rst [hunk 3]
```diff
@@ -737,6 +744,44 @@ section.
...
+.. code:: python
+
+   # ------------------------------------------
+   # Instrument test attributes
+
+   _clean_warn = {inst_id: {tag: {'dusty': [
+                                ('logger', 'WARN', "I am a warning!", 'clean'),
+                                ('warning', UserWarning,
+				 'I am a serios warning!', 'dusty'),
+                                ('error, ValueError, 'I am an error', 'dusty')]}
```
Comment (aburrell) [inline docs/new_instrument.rst:778]:
```suggestion
                                ('error', ValueError, 'I am an error', 'dusty')]}
```

**Review by jklenzing — COMMENTED** [review]
Looks good. I would suggest restructuring so that tests where nothing is run are skipped to be consistent with similar tests elsewhere.

**Comment by aburrell** [comment]
The decrease in coverage is consistent with tests not being skipped and edge cases for ecosystem packages not being triggered.

File: pysat/tests/classes/cls_instrument_library.py [hunk 4]
```diff
@@ -66,14 +66,57 @@ def initialize_test_inst_and_date(inst_dict):
...
-                                 **kwargs)
+                                 temporary_file_list=True, update_files=True,
+                                 use_header=True, **kwargs)
     test_dates = inst_dict['inst_module']._test_dates
     date = test_dates[inst_dict['inst_id']][inst_dict['tag']]
     return test_inst, date
 
 
+def set_strict_time_flag(test_inst, date, raise_error=False):
+    """Ensure the strict time flag does not interfere with other tests.
```
Comment (jklenzing) [inline pysat/tests/classes/cls_instrument_library.py:77]: Technically, this loads data in addition to setting the `strict_time_flag`. Tested this code manually, and it does what it should. Modifying `test_inst` in place makes me a little nervous, but it's probably fine as long as the docstring is clear.

File: pysat/tests/classes/cls_instrument_library.py [hunk 5]
```diff
@@ -339,20 +384,10 @@ def test_load(self, clean_level, inst_dict):
...
-                        test_inst.load(date=date, use_header=True)
-                    assert len(war) >= 1
-                    categories = [war[j].category for j in range(0, len(war))]
-                    assert UserWarning in categories
-                else:
-                    # If error message does not match, raise error anyway
-                    raise(verr)
+
+            # Make sure the strict time flag doesn't interfere with
+            # the cleaning tests
```
Comment (jklenzing) [inline pysat/tests/classes/cls_instrument_library.py:389]:
```suggestion
            # Make sure the strict time flag doesn't interfere with
            # the load tests
```

File: pysat/tests/classes/cls_instrument_library.py [hunk 6]
```diff
@@ -339,20 +384,10 @@ def test_load(self, clean_level, inst_dict):
...
-                else:
-                    # If error message does not match, raise error anyway
-                    raise(verr)
+
+            # Make sure the strict time flag doesn't interfere with
+            # the load tests
+            set_strict_time_flag(test_inst, date, raise_error=True)
 
             # Make sure fake data is cleared
             assert target not in test_inst.data
```
Comment (jklenzing) [inline pysat/tests/classes/cls_instrument_library.py:393]: this is the bit that has changed. `set_strict_time_flag` loads the data (which is why 393 passes). But in some cases the new function will force clean to be 'none', which is a change in behaviour.

File: pysat/tests/classes/cls_instrument_library.py [hunk 7]
```diff
@@ -66,14 +66,64 @@ def initialize_test_inst_and_date(inst_dict):
...
-                                 update_files=True, use_header=True,
-                                 **kwargs)
+                                 temporary_file_list=True, update_files=True,
+                                 use_header=True, **kwargs)
     test_dates = inst_dict['inst_module']._test_dates
     date = test_dates[inst_dict['inst_id']][inst_dict['tag']]
     return test_inst, date
 
 
+def set_strict_time_flag(test_inst, date, raise_error=False, clean_off=True):
```
Comment (jklenzing) [inline pysat/tests/classes/cls_instrument_library.py:76]:
```suggestion
def load_and_set_strict_time_flag(test_inst, date, raise_error=False, clean_off=True):
```

File: pysat/tests/classes/cls_instrument_library.py [hunk 8]
```diff
@@ -66,14 +66,64 @@ def initialize_test_inst_and_date(inst_dict):
...
+    """Ensure the strict time flag does not interfere with other tests.
```
Comment (jklenzing) [inline pysat/tests/classes/cls_instrument_library.py:77]:
```suggestion
    """Load data and set the strict time flag if needed for other tests.
```

File: pysat/tests/classes/cls_instrument_library.py [hunk 9]
```diff
@@ -366,6 +409,108 @@ def test_load(self, clean_level, inst_dict):
...
+            # Cleaning warnings may vary by clean level, test the warning
+            # messages at the current clean level, specified by `clean_level`
+            if clean_level in clean_warn.keys():
+                # Only need to test if there are clean warnings for this level
+                test_inst, date = initialize_test_inst_and_date(inst_dict)
+                clean_warnings = clean_warn[clean_level]
+
+                # Make sure the strict time flag doesn't interfere with
+                # the cleaning tests
+                set_strict_time_flag(test_inst, date)
```
Comment (jklenzing) [inline pysat/tests/classes/cls_instrument_library.py:446]:
```suggestion
                load_and_set_strict_time_flag(test_inst, date)
```

File: pysat/tests/classes/cls_instrument_library.py [hunk 10]
```diff
@@ -335,24 +387,15 @@ def test_load(self, clean_level, inst_dict):
...
-                    categories = [war[j].category for j in range(0, len(war))]
-                    assert UserWarning in categories
-                else:
-                    # If error message does not match, raise error anyway
-                    raise(verr)
+
+            # Make sure the strict time flag doesn't interfere with
+            # the load tests, and re-run with desired clean level
+            set_strict_time_flag(test_inst, date, raise_error=True,
+                                 clean_off=False)
```
Comment (jklenzing) [inline pysat/tests/classes/cls_instrument_library.py:399]:
```suggestion
            load_and_set_strict_time_flag(test_inst, date, raise_error=True,
                                          clean_off=False)

```

**Review by jklenzing — CHANGES_REQUESTED** [review]

**Review by jklenzing — APPROVED** [review]
